tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -25,25 +29,189 @@ pub struct CudaVec<T: Numeric> {
+    pub unsafe fn new_async(len: usize, stream: &CudaStream) -> Self {
+        let size = len as u64 * std::mem::size_of::<T>() as u64;
+        let ptr = CudaPtr {
+            ptr: cuda_malloc_async(size, stream.as_c_ptr()),
+            device: stream.device,
+        };
+
         Self {
             ptr,
             len,
-            device,
+            device: stream.device,
             _phantom: PhantomData,
         }
     }
+
     /// # Safety
     ///
     /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
     ///   not be dropped until stream is synchronised
-    pub unsafe fn from_async(h_data: &Vec<T>, stream: &CudaStream) -> Self {
-        let mut d_data = stream.malloc_async(h_data.len() as u32);
-        stream.copy_to_gpu_async(&mut d_data, h_data.as_slice());
-        d_data
+    pub unsafe fn from_async(src: &[T], stream: &CudaStream) -> Self {
+        let mut res = Self::new_async(src.len(), stream);
+        // We have to check that h_data is not empty, because cuda_memset with size 0 is invalid
+        if !src.is_empty() {
+            res.copy_to_gpu_async(src, stream);
+        }
+        res
+    }
+
+    /// # Safety
+    ///
+    /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
+    ///   not be dropped until stream is synchronised
+    pub unsafe fn memset_async(&mut self, value: T, stream: &CudaStream)
+    where
+        T: Into<u64>,
+    {
+        let size = self.len() * std::mem::size_of::<T>();
+        // We check that self is not empty to avoid invalid pointers
+        if size > 0 {
+            cuda_memset_async(
+                self.as_mut_c_ptr(),
+                value.into(),
+                size as u64,
+                stream.as_c_ptr(),
+            );
+        }
+    }
+
+    /// Copies data from slice into `CudaVec`
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_to_gpu_async(&mut self, src: &[T], stream: &CudaStream)
+    where
+        T: Numeric,
+    {
+        let src_size = std::mem::size_of_val(src);
+        assert!(self.len() * std::mem::size_of::<T>() >= src_size);
+
+        // We have to check that src is not empty, because Rust slice with size 0 results in an
+        // invalid pointer being passed to copy_to_gpu_async
+        if src_size > 0 {
+            cuda_memcpy_async_to_gpu(
+                self.as_mut_c_ptr(),
+                src.as_ptr().cast(),
+                src_size as u64,
+                stream.as_c_ptr(),
+            );
+        }
+    }
+
+    /// Copies data between two `CudaVec`
+    ///
+    /// # Safety
+    ///
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_gpu_to_gpu_async(&mut self, src: &Self, stream: &CudaStream)
```

Comment:
The name is `copy_to_gpu_async`, but since ,its  `&mut self`, is should be called  `copy_from_gpu_async`